src/consensus/replay_tower.zig

```diff
@@ -1579,6 +1613,622 @@ fn optimisticallyBypassVoteStakeThresholdCheck(
+            &vote_account,
+        );
+
+        for (vote_state.votes.constSlice()) |vote| {
+            const interval = try lockout_intervals
+                .getOrPut(allocator, vote.lastLockedOutSlot());
+            if (!interval.found_existing) {
+                interval.value_ptr.* = std.ArrayList(VotedSlotAndPubkey).init(allocator);
+            }
+            try interval.value_ptr.*.append(.{ .slot = vote.slot, .pubkey = key });
+        }
+
+        // Vote account for this validator
+        if (key.equals(vote_account_pubkey)) {
+            my_latest_landed_vote = if (vote_state.nthRecentLockout(0)) |l| l.slot else null;
+            logger.debug().logf("vote state {any}", .{vote_state});
+            const observed_slot = if (vote_state.nthRecentLockout(0)) |l| l.slot else 0;
+
+            logger.debug().logf("observed slot {any}", .{observed_slot});
+        }
+        const start_root = vote_state.root_slot;
+
+        // Add the last vote to update the `heaviest_subtree_fork_choice`
+        if (vote_state.lastVotedSlot()) |last_landed_voted_slot| {
+            if (progress_map.getHash(last_landed_voted_slot)) |frozen_hash| {
+                _ = try latest_validator_votes_for_frozen_banks.checkAddVote(
+                    allocator,
+                    key,
+                    last_landed_voted_slot,
+                    frozen_hash,
+                    .replay,
+                );
+            }
+        }
+
+        // Simulate next vote and extract vote slots using the provided bank slot.
+        try vote_state.processNextVoteSlot(bank_slot);
+
+        for (vote_state.votes.constSlice()) |vote| {
+            try vote_slots.put(vote.slot);
+        }
+
+        if (start_root != vote_state.root_slot) {
+            if (start_root) |root| {
+                const vote = Lockout{ .slot = root, .confirmation_count = MAX_LOCKOUT_HISTORY };
+                logger.trace().logf("ROOT: {}", .{vote.slot});
+                try vote_slots.put(vote.slot);
+            }
+        }
+        if (vote_state.root_slot) |root| {
+            const vote = Lockout{ .slot = root, .confirmation_count = MAX_LOCKOUT_HISTORY };
+            try vote_slots.put(vote.slot);
+        }
+
+        // The last vote in the vote stack is a simulated vote on bank_slot, which
+        // we added to the vote stack earlier in this function by calling processVote().
+        // We don't want to update the ancestors stakes of this vote b/c it does not
+        // represent an actual vote by the validator.
+
+        // Note: It should not be possible for any vote state in this bank to have
+        // a vote for a slot >= bank_slot, so we are guaranteed that the last vote in
+        // this vote stack is the simulated vote, so this fetch should be sufficient
+        // to find the last unsimulated vote.
+        std.debug.assert(
+            if (vote_state.nthRecentLockout(0)) |l| l.slot == bank_slot else false,
+        );
```

Comment:
I think this assertion would be clearer like this:
```suggestion
        std.debug.assert(vote_state.nthRecentLockout(0).?.slot == bank_slot);
```
That way if `nthRecentLockout` returns null, it trips the null unwrap; otherwise, it will trip the assert - thus distinguishing these two illegal invariants.

Reply:
[5f62fa2](https://github.com/Syndica/sig/pull/798/commits/5f62fa2a26303a876bcec1b00d803f9fddf2da43)